keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -0,0 +1,2314 @@
+  class for all containers
+- `rdt.resources.kubernetes.io/container.<container-name>` for
+  container-specific RDT class settings
+  blockio class for all containers
+- `blockio.resources.kubernetes.io/default` for setting a Pod-level default
+- `blockio.resources.kubernetes.io/container.<container-name>` for
+  container-specific blockio class settings
+
+#### Kubelet
+
+Kubelet would interpret the specific [pod annotations](#pod-annotations) and
+translate them into corresponding `QOSResources` data in the CRI
+ContainerConfig message at container creation time (CreateContainerRequest).
+Pod-level QoS-class would not supported at this point (via pod annotations).
+
+A feature gate `TranslateQoSPodMetadata` would enable kubelet to interpretthe
+specific pod annotations. If the feature gate is disabled the annotations would
+simply be ignored by kubelet.
+
+#### API server
+
+A validation check (core api validation) would added in the API server to
+reject changes to the QoS-class resource specific
+[pod annotations](#pod-annotations) after a Pod has been created. This would
+ensure that the annotations always reflect the actual assignment of QoS-class
+resources of a Pod. It also would serve as part of the UX to indicate the
+in-place updates of the resources via annotations is not supported.
+
+### Class capacity with extended resources
+
+Support for class capacity could be left out of the proposal to simplify the
+concept. It would be possible to implement "class capacity" by leveraging
+extended resources and mutating admission webhooks:
+
+1. An extended resource with the desited capacity is created for each class
+   which needs to be controlled. A possible example:
+  ```plaintext
+  Allocatable:
+  qos/container/resource-X/class-A:        2
+  qos/container/resource-X/class-B:        4
+  qos/container/resource-Y/class-foo:      1
+  ```
+
+  In this example *class-A* and *class-B* of QoS-class resource *resource-X*
+  would be limited to 2 and 4 assignments, respectively (on this node).
+  Similarly, class *class-foo* of QoS-class resource *resource-Y* would be
+  limited to only one assignment.
+
+2. A specific mutating admission webhook is deployed which syncs the specific
+   extended resources. That is, if a specific QoS-class resource assignment is
+   requested in the pod spec, a request for the corresponding extended resource
+   is automatically added. And vice versa: if a specific extended resource is
+   requested a QoS-class resource request is automatically put in place.
+
+However, this approach is quite involved, has shortcomings and caveats and thus
+it is probably suitable only for targeted usage scenarios, not as a general
+solution. Downsides include:
+
+- requires implementation of "side channel" control mechanisms, e.g. admission
+  webhook and some solution for capacity management (extended resources)
+- deployment of admission webhooks is cumbersome
```

Comment:
Validation admission policies mitigates this somewhat.

Reply:
But can you do mutating webhooks with that?

Reply:
It's actually in the works.

Reply:
Noted. I added mention about this in the KEP